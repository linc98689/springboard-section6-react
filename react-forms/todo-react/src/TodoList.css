.todolist{
    color:rgb(28, 21, 53);
    background-color:rgb(230, 246, 246);
}
.todolist-title{
    text-align:center;
}

.todolist-subtitle{
    font-size: 1.2em;
}