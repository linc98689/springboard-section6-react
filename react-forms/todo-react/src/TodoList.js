import React, {useState, useEffect} from "react";
import "./TodoList.css";
import NewTodoForm from "./NewTodoForm";
import Todo from "./Todo";
import EditForm from "./EditForm";

let stored = localStorage.getItem("todo-list");
let initList = stored ===null? []: JSON.parse(stored);
const TodoList = ()=>{
    const [tasks, setTasks] = useState(initList);
    const [currentTaskIdx, setCurrentTaskIdx] = useState(-1); //not editing
    useEffect(()=>{
        localStorage.setItem("todo-list", JSON.stringify(tasks));
    }, [tasks])

    const addTask = (task)=>{
        setTasks((data) =>[{task, isdone:false}, ...data]);
    };

    const removeTask = (idx)=>{
        setTasks((data) => data.toSpliced(idx, 1))
    }

    const updateTaskStatus = (idx)=>{
        let todo = tasks[idx];
        todo.isdone = !todo.isdone;
        setTasks(data => data.toSpliced(idx, 1, todo));
    }

    const updateTask = (task)=>{
        const todo = tasks[currentTaskIdx];
        todo.task = task;
        setTasks(data => data.toSpliced(currentTaskIdx, 1, todo));
    }

    const updateCurrentTaskIdx = (idx)=>{
        setCurrentTaskIdx(data => idx);
    }

    //return
    return (
        <div className="todolist">
            <h1 className="todolist-title">My Todo List</h1>
            <h2 className="todolist-subtitle">Add a new task</h2>
            <NewTodoForm addTask={addTask}/>

            <h2 className="todolist-subtitle"> Tasks on hand</h2>
            <ul className="todolist-list">
                {tasks.map((e, i)=> {
                    return (<Todo id={i} task={e.task} isdone={`${e.isdone}`} 
                        key={i}
                        removeTask={removeTask}
                        updateTaskStatus={updateTaskStatus} 
                        updateCurrentTaskIdx={updateCurrentTaskIdx}/>)
                })}
            </ul>

            <EditForm  updateTask={updateTask}
             updateCurrentTaskIdx={updateCurrentTaskIdx}
             task={currentTaskIdx === -1? "": tasks[currentTaskIdx].task}
            className_2={currentTaskIdx === -1?
             "edit-form-unshown" : "edit-form-shown"}/>
        </div>
    );
}

export default TodoList