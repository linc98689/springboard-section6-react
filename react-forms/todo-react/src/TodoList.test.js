import { render } from '@testing-library/react';
import '@testing-library/jest-dom';
import TodoList from "./TodoList";

// smoke test
test("should render", ()=>{
    render(<TodoList/>)
});

// snapshot test
test("should not change", ()=>{
    const {asFragment} = render(<TodoList />);
    expect(asFragment()).toMatchSnapshot();
})
