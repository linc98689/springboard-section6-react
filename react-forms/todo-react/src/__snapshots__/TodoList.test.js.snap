// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should not change 1`] = `
<DocumentFragment>
  <div
    class="todolist"
  >
    <h1
      class="todolist-title"
    >
      My Todo List
    </h1>
    <h2
      class="todolist-subtitle"
    >
      Add a new task
    </h2>
    <form>
      <input
        class="newtask-input"
        placeholder="new task"
        value=""
      />
      <button
        class="newtask-btn"
        type="submit"
      >
        Add
      </button>
    </form>
    <h2
      class="todolist-subtitle"
    >
       Tasks on hand
    </h2>
    <ul
      class="todolist-list"
    />
    <div
      class="edit-form edit-form-unshown"
    >
      <h2
        class="edit-form-title"
      >
         Editing a task
      </h2>
      <form>
        <input
          class="edit-input"
          id="edit-task"
          type="text"
          value=""
        />
        <div
          class="edit-buttons"
        >
          <button
            type="button"
          >
            Cancel
          </button>
          <button
            type="submit"
          >
            Save
          </button>
        </div>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
